### ISSUE TITLE
noise on the specific frequency band when do synthesis

### ISSUE THREAD

**Issue by hubeibei007** [NVIDIA/tacotron2#142]
Hi,

I have trained the model on a Mandarin dataset. Currently it achieved the coverage and get a linear alignment. But I found there are noises on the specific frequency band, such as 5.5khz, 2.75khz, 8.5khz. The following is the synthesis spectrogram. When listen the generated audio samples, the sound is not clean enough. Could anyone share your opinion about the problem? Why my spectrogram looks like that?

Thanks a lot!

![211550122469_ pic_hd](https://user-images.githubusercontent.com/17315133/52765043-708ea880-305d-11e9-9d54-cee136298652.jpg)

**Comment by rafaelvalle** [comment]
Did you train waveglow on your Mandarin dataset?

**Comment by hubeibei007** [comment]
Hi Rafael,

Thanks a lot for your reply. I trained the tacotron2 on my Mandarin dataset, and finetuning the waveglow  model on the same Mandarin dataset. During the synthesis procedure, the mel spectrograms of each step are the followings:

The inferenced melspec of tacotron2:
![wechatimg25](https://user-images.githubusercontent.com/17315133/53220590-4c0b7f80-36a0-11e9-91de-63ae56f2e78e.jpeg)

Then the audio was generated by my waveglow model. The following is the melspec of the generated audio. It seems that there are horizontal lines in this melspec. And the voice is not clean.
![wechatimg26](https://user-images.githubusercontent.com/17315133/53220617-807f3b80-36a0-11e9-9de2-b61f22ff73f6.jpeg)

The audio parameters of tacotron2 and waveglow are same. The figures were ploted by the code of this project. The loss of waveglow is about -5.4 and the loss of tacotron2 is 0.26. 

Could you please share your views on my problem? Is the herizontal lines in the final melspec is normal?  Should I continue to train the waveglow to get a better convergence? What should I do to improve the audio quality based on my dataset?

Thanks a lot!

**Comment by rafaelvalle** [comment]
The horizontal lines are waveglow's bias. They should go away as the likelihood improves.
A quick hack is to subtract the noise bias from the noisy audio on the frequency domain. Here's an notebook example, where noise.wav is a slice of audio with waveglow bias noise only and noisy_audio is some other audio generated with waveglow.

```
%matplotlib inline
import IPython.display as ipd               
import matplotlib
import matplotlib.pylab as plt
import torch
from torch.autograd import Variable
import librosa
from layers import STFT


stft = STFT(filter_length=2048, hop_length=1024, win_length=2048)
noise, sr_noise = librosa.load("/home/scratch.adlr-gcf/rafaelvalle/noise.wav", sr=None)
audio, sr_audio = librosa.load("/home/scratch.adlr-gcf/rafaelvalle/noisy_audio.wav", sr=None)

noise_t = torch.autograd.Variable(torch.from_numpy(noise))[None, :]
audio_t = torch.autograd.Variable(torch.from_numpy(audio))[None, :]

noise_spec, noise_angles = stft.transform(noise_t)
audio_spec, audio_angles = stft.transform(audio_t)

scale = 2.5
sub_frames = noise_spec[:, :, 0][:, :, None] * scale
audio_spec_denoised = torch.clamp(audio_spec - sub_frames, 0.0)
audio_denoised = stft.inverse(audio_spec_denoised, audio_angles)

fig, axes = plt.subplots(4, 1, figsize=(16, 12))
axes[0].imshow(librosa.core.amplitude_to_db(noise_spec[0].data.cpu().numpy()), aspect='auto', interpolation='none', origin='bottom')
axes[1].imshow(librosa.core.amplitude_to_db(audio_spec[0].data.cpu().numpy()), aspect='auto', interpolation='none', origin='bottom')
axes[2].imshow(librosa.core.amplitude_to_db(audio_spec_denoised[0].data.cpu().numpy()), aspect='auto', interpolation='none', origin='bottom')
axes[3].imshow(librosa.core.amplitude_to_db(sub_frames[0].data.cpu().numpy()), aspect='auto', interpolation='none', origin='bottom')
plt.tight_layout()
ipd.Audio(audio, rate=sr_audio)
ipd.Audio(audio_denoised[0].data.cpu().numpy(), rate=sr_audio) 
```

**Comment by rafaelvalle** [comment]
You can improve Tacotron2 validation loss by annealing the learning rate and training longer.

**Comment by hubeibei007** [comment]
The method of subtraction the noise bias from my generated audio improves my audio quality. 
I will try to improve Tacotron2 validation. 
Thank you very much.

**Comment by yxt132** [comment]
> stft = STFT(filter_length=filter_length, hop_length=hop_length, win_length=win_length)
> 
> noise, sr_noise = librosa.load("noise.wav", sr=None)
> audio, sr_audio = librosa.load("noisy_audio.wav", sr=None)
> 
> noise_spec, noise_angles = stft.transform(torch.autograd.Variable(torch.from_numpy(noise))[None, :])
> audio_spec, audio_angles = stft.transform(torch.autograd.Variable(torch.from_numpy(audio))[None, :])
> 
> strength = 16
> sub_frames = noise_spec[:, :, 0][:, :, None] * strength
> audio_spec_denoised = torch.clamp(audio_spec - sub_frames, 0.0)
> audio_denoised = stft.inverse(audio_spec_denoised, audio_angles)

@rafaelvalle Is there a better way to get rid of the horizontal lines during training rather than doing post processing like this?

**Comment by anupam456** [comment]
Hi @hubeibei007 

As mentioned below in your response, how did you generate the noise.wav file?

> The method of subtraction the noise bias from my generated audio improves my audio quality.
> I will try to improve Tacotron2 validation.
> Thank you very much.

**Comment by hubeibei007** [comment]
Hi @anupam456 

I chose some non voice segments from the generated audio as the noise input.

**Comment by rafaelvalle** [comment]
The code to remove the model bias, aka line noise, has been added here:
https://github.com/NVIDIA/tacotron2/blob/master/inference.ipynb

**Comment by yxt132** [comment]
> The code to remove the model bias, aka line noise, has been added here:
> https://github.com/NVIDIA/tacotron2/blob/master/inference.ipynb
> It only requires the waveglow model: not need to save an audio file with the model bias/noise

I just tried but could not find the denoiser module.

**Comment by rafaelvalle** [comment]
Try updating the submodule:
`git submodule update --remote --merge`
